server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -560,18 +560,24 @@ public DocumentMapper merge(String type, CompressedXContent mappingSource, Merge
         return doMerge(type, reason, mappingSourceAsMap);
     }
 
-    private synchronized DocumentMapper doMerge(String type, MergeReason reason, Map<String, Object> mappingSourceAsMap) {
+    private DocumentMapper doMerge(String type, MergeReason reason, Map<String, Object> mappingSourceAsMap) {
         Mapping incomingMapping = parseMapping(type, reason, mappingSourceAsMap);
-        Mapping mapping = mergeMappings(this.mapper, incomingMapping, reason, this.indexSettings);
-        // TODO: In many cases the source here is equal to mappingSource so we need not serialize again.
-        // We should identify these cases reliably and save expensive serialization here
-        DocumentMapper newMapper = newDocumentMapper(mapping, reason, mapping.toCompressedXContent());
         if (reason == MergeReason.MAPPING_AUTO_UPDATE_PREFLIGHT) {
-            return newMapper;
+            Mapping mapping = mergeMappings(this.mapper, incomingMapping, MergeReason.MAPPING_AUTO_UPDATE_PREFLIGHT, this.indexSettings);
+            // TODO: In many cases the source here is equal to mappingSource so we need not serialize again.
+            // We should identify these cases reliably and save expensive serialization here
+            return newDocumentMapper(mapping, MergeReason.MAPPING_AUTO_UPDATE_PREFLIGHT, mapping.toCompressedXContent());
+        } else {
+            synchronized (this) {
+                Mapping mapping = mergeMappings(this.mapper, incomingMapping, reason, this.indexSettings);
```

Comment:
Maybe de-duplicate the TODO given it's the same, and leave a clarifying comment on the distinction and why mergeMappings needs to be under synchronized in one case but not on the other.

Reply:
++ makes sense, added